### ISSUE TITLE
markdown-imenu-create-nested-index function definition is void

### ISSUE THREAD

**Issue by finger563** [ardumont/markdown-toc#29]
Hello,

Please, provide the following information and remove the useless parts.
Thanks in advance.
## bug?

Yes, 

markdown-imenu-create-nested-index, find-library-name are not known to be defined, produce errors when running `markdown-toc-generate-toc`.
### M-x markdown-toc-bug-report

Report back the output from that command here.
(You can quote it with ``` for a better readability).
### Expected behavior

What command did you use and what were you expecting?

A TOC generated when running `markdown-toc-generate-toc`
### Actual behavior

What really happened?

error, `Symbol's function definition is void: markdown-imenu-create-nested-index`
### Steps to reproduce the behavior

How can we try and reproduce?
### Why is it better?

Thanks for sharing!

Cheers,

**Comment by ardumont** [comment]
Hello,

Can you please:

- `M-x markdown-toc-bug-report` and report back the result here (the content is in the `*Messages*` buffer, so just copy/paste here)? Thanks.
- clean up your bug description by removing the unneeded entries (the one you did not fill in)?

This is possibly:
- related to the closed bugs #20, #22. Have you checked?
- due to an old markdown-mode version (so the fix would possibly be to upgrade it).

Cheers,

**Comment by anarcat** [comment]
i'm not the original submitter but i am seeing the same issue.

I did this:

* installed the package through marmalade (`M-x package-list RET`, find the package, `i x`...)
* switched to an existing markdown buffer
* ran `M-x markdown-toc-generate-toc`
* Got: `markdown-toc-generate-toc: Symbol’s function definition is void: markdown-imenu-create-nested-index`
* Expected: a TOC. :)

System information:
- system-type: gnu/linux
- locale-coding-system: utf-8-unix
- emacs-version: GNU Emacs 25.1.1 (x86_64-pc-linux-gnu, GTK+ Version 3.22.5)
 of 2016-12-31, modified by Debian
- markdown-toc version: 0.1.2
- markdown-toc path: /home/anarcat/.emacs.d/elpa/markdown-toc-20170224.2350/markdown-toc.el

This is using markdown-mode 2.1 from the [Debian packages](https://tracker.debian.org/pkg/markdown-mode) which dates from about february 2016. Upgrading to the snapshot version available in marmalade (`20161222.1416`) fixes the issue. It would be nice to have a way for the toc plugin to handle this issue more gracefully than just crash - maybe warn the user they are using an old version?

Alternatively, couldn't some backwards-compatibility magic be done to backport that missing code into the plugin itself?

**Comment by ardumont** [comment]
Hello,

> This is using markdown-mode 2.1 from the Debian packages which dates from about february 2016. Upgrading to the snapshot version available in marmalade (20161222.1416) fixes the issue.

Thank you, now i'm able to reproduce the issue (i am running debian as well).
(Note that it could also be a more recent from melpa).

I have updated M-x markdown-toc-bug-report to add the library path to markdown-mode, this apparently can help :D 

> It would be nice to have a way for the toc plugin to handle this issue more gracefully than just crash 

Indeed.

> maybe warn the user they are using an old version?

Now that i understand it's indeed the case why not.
But i'd be more inclined to solve it now.

> Alternatively, couldn't some backwards-compatibility magic be done to backport that missing code into the plugin itself?

Mmm, maybe.

This is the commit that introduced the issue for your system version.
https://github.com/ardumont/markdown-toc/commit/18e4d7910608ac86a1fa0dfdbca7de2a06ec22cb

If you eval this snippet (<kbd>M-: `paste-the-snippet-below` RET</kbd>):

```lisp
(defun markdown-toc-generate-toc (&optional replace-toc-p)
  "Generate a TOC for markdown file at current point.
Deletes any previous TOC.
If called interactively with prefix arg REPLACE-TOC-P, replaces previous TOC."
  (interactive "P")
  (save-excursion
    (when (markdown-toc--toc-already-present-p)
      ;; when toc already present, remove it
      (let ((region-start (markdown-toc--toc-start))
            (region-end   (markdown-toc--toc-end)))
        (delete-region region-start (1+ region-end))
        (when replace-toc-p
          (goto-char region-start))))
    (->> (markdown-imenu-create-index)
         markdown-toc--compute-toc-structure
         (funcall markdown-toc-user-toc-structure-manipulation-fn)
         markdown-toc--generate-toc
         insert)))
```
And now trying to generate the toc again, everything should be fine. Am i right?
(It's the case for me).

--

Assuming it is.
Please, be a little more patient with me, can you please try this snippet again?

```lisp
(defun markdown-toc-generate-toc (&optional replace-toc-p)
  "Generate a TOC for markdown file at current point.
Deletes any previous TOC.
If called interactively with prefix arg REPLACE-TOC-P, replaces previous TOC."
  (interactive "P")
  (save-excursion
    (when (markdown-toc--toc-already-present-p)
      ;; when toc already present, remove it
      (let ((region-start (markdown-toc--toc-start))
            (region-end   (markdown-toc--toc-end)))
        (delete-region region-start (1+ region-end))
        (when replace-toc-p
          (goto-char region-start))))
    (->> (funcall imenu-create-index-function)
         markdown-toc--compute-toc-structure
         (funcall markdown-toc-user-toc-structure-manipulation-fn)
         markdown-toc--generate-toc
         insert)))
```

The last snippet should be working as well but with the potential advantage to work with both version...
markdown-mode uses a defvar to store the index function to use to generate the index since at least that markdown-mode 2.1 version. So this should be forward/backward compatible (well, depending on your point of view :).

Cheers,

**Comment by anarcat** [comment]
On 2017-02-27 10:04:06, Antoine R. Dumont wrote:
> Hello,
>
> If you eval this snippet (<kbd>M-: `paste-the-snippet-below` RET</kbd>):
>
> ```lisp
> (defun markdown-toc-generate-toc (&optional replace-toc-p)
>   "Generate a TOC for markdown file at current point.
> Deletes any previous TOC.
> If called interactively with prefix arg REPLACE-TOC-P, replaces previous TOC."
>   (interactive "P")
>   (save-excursion
>     (when (markdown-toc--toc-already-present-p)
>       ;; when toc already present, remove it
>       (let ((region-start (markdown-toc--toc-start))
>             (region-end   (markdown-toc--toc-end)))
>         (delete-region region-start (1+ region-end))
>         (when replace-toc-p
>           (goto-char region-start))))
>     (->> (markdown-imenu-create-index)
>          markdown-toc--compute-toc-structure
>          (funcall markdown-toc-user-toc-structure-manipulation-fn)
>          markdown-toc--generate-toc
>          insert)))
> ```
> And now trying to generate the toc again, everything should be fine. Am i right?
> (It's the case for me).

Yes, it works.

> --
>
> Assuming it is.
> Please, be a little more patient with me, can you please try this snippet again?
>
> ```lisp
> (defun markdown-toc-generate-toc (&optional replace-toc-p)
>   "Generate a TOC for markdown file at current point.
> Deletes any previous TOC.
> If called interactively with prefix arg REPLACE-TOC-P, replaces previous TOC."
>   (interactive "P")
>   (save-excursion
>     (when (markdown-toc--toc-already-present-p)
>       ;; when toc already present, remove it
>       (let ((region-start (markdown-toc--toc-start))
>             (region-end   (markdown-toc--toc-end)))
>         (delete-region region-start (1+ region-end))
>         (when replace-toc-p
>           (goto-char region-start))))
>     (->> (funcall imenu-create-index-function)
>          markdown-toc--compute-toc-structure
>          (funcall markdown-toc-user-toc-structure-manipulation-fn)
>          markdown-toc--generate-toc
>          insert)))
> ```
>
> The last snippet should be working as well but with the potential
> advantage to work with both version...

It works as well.

> markdown-mode uses a defvar to store the index function to use to generate the index since at least that markdown-mode 2.1 version. So this should be forward/backward compatible (well, depending on your point of view :).

Great! Can't wait to see this hit the mirrors. :)

A.

-- 
Worker bees can leave
Even drones can fly away
The queeen is their slave
                        - Fight Club
